doc/man/man3/Xcm.3
.TH "X Color Management Core API's" 3 "Tue May 26 2015" "Version 0.5.4" "Xcm-0.5.4" \" -*- nroff -*-
.ad l
.nh
.SH NAME
X Color Management Core API's \- 
.SS "Data Structures"

.in +1c
.ti -1c
.RI "struct \fBXcolorProfile\fP"
.br
.ti -1c
.RI "struct \fBXcolorRegion\fP"
.br
.in -1c
.SS "Macros"

.in +1c
.ti -1c
.RI "#define \fBXCM_COLOR_PROFILES\fP   '_ICC_COLOR_PROFILES'"
.br
.ti -1c
.RI "#define \fBXCM_COLOR_REGIONS\fP   '_ICC_COLOR_REGIONS'"
.br
.ti -1c
.RI "#define \fBXCM_COLOR_OUTPUTS\fP   '_ICC_COLOR_OUTPUTS'"
.br
.ti -1c
.RI "#define \fBXCM_COLOR_DESKTOP\fP   '_ICC_COLOR_DESKTOP'"
.br
.ti -1c
.RI "#define \fBXCM_DEVICE_PROFILE\fP   '_ICC_DEVICE_PROFILE'"
.br
.in -1c
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum { \fBXCM_COLOR_SERVER_REGIONS\fP = 0x01, \fBXCM_COLOR_SERVER_PROFILES\fP = 0x02, \fBXCM_COLOR_SERVER_DISPLAY_ADVANCED\fP = 0x04, \fBXCM_COLOR_SERVER_OUTPUTS\fP = 0x08, \fBXCM_COLOR_SERVER_MANAGEMENT\fP = 0x10 }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "int \fBXcolorProfileUpload\fP (Display *dpy, \fBXcolorProfile\fP *profile)"
.br
.RI "\fIUploads the profile into all screens of the display\&. \fP"
.ti -1c
.RI "int \fBXcolorProfileDelete\fP (Display *dpy, \fBXcolorProfile\fP *profile)"
.br
.RI "\fIDecreases the ref-count of the profile\&. \fP"
.ti -1c
.RI "int \fBXcolorRegionInsert\fP (Display *dpy, Window win, unsigned long pos, \fBXcolorRegion\fP *region, unsigned long nRegions)"
.br
.RI "\fIInserts the supplied regions into the stack\&. \fP"
.ti -1c
.RI "\fBXcolorRegion\fP * \fBXcolorRegionFetch\fP (Display *dpy, Window win, unsigned long *nRegions)"
.br
.RI "\fIFetches the existing regions\&. \fP"
.ti -1c
.RI "int \fBXcolorRegionDelete\fP (Display *dpy, Window win, unsigned long start, unsigned long count)"
.br
.RI "\fIDeletes regions\&. \fP"
.ti -1c
.RI "int \fBXcolorRegionActivate\fP (Display *dpy, Window win, unsigned long start, unsigned long count)"
.br
.RI "\fIActivates regions\&. \fP"
.ti -1c
.RI "int \fBXcmColorServerCapabilities\fP (Display *dpy)"
.br
.RI "\fIinforms which colour server services are available \fP"
.in -1c
.SH "Detailed Description"
.PP 

.SH "Macro Definition Documentation"
.PP 
.SS "#define XCM_COLOR_DESKTOP   '_ICC_COLOR_DESKTOP'"
The XCM_COLOR_DESKTOP macro The atom is attached on the root window to tell about a color servers activity\&. The content is of type XA_STRING and has four sections separated by a empty space char ' '\&. The _ICC_COLOR_DESKTOP atom is a string with following usages:
.IP "\(bu" 2
uniquely identify the colour server
.IP "\(bu" 2
tell the name of the colour server
.IP "\(bu" 2
tell the colour server is alive All sections are separated by one space char ' ' for easy parsing\&.
.PP
.PP
The first section contains the process id (pid_t) of the color server process, which has set the atom\&. The second section contains time since epoch GMT as returned by time(NULL)\&. The thired section contains the bar '|' separated and surrounded capabilities:
.IP "\(bu" 2
ICP _ICC_COLOR_PROFILES - support per region profiles
.IP "\(bu" 2
ICM _ICC_COLOR_MANAGEMENT - color server is active
.IP "\(bu" 2
ICO _ICC_COLOR_OUTPUTS - support per window and output configuration
.IP "\(bu" 2
ICR _ICC_COLOR_REGIONS - support regions
.IP "\(bu" 2
ICA _ICC_COLOR_DISPLAY_ADVANCED - use CMS advanced settings, e\&.g\&. proofing
.IP "\(bu" 2
V0\&.3 indicates version compliance to the _ICC_Profile in X spec The fourth section contains the servers name identifier\&.
.PP
.PP
As of this specification the third section must contain ICR and the supported _ICC_PROFILE in X version\&.
.PP
A example of a valid atom might look like: _ICC_COLOR_DESKTOP(STRING) = '4518 1274001512 |ICR|ICM|V0\&.3| compiz_colour_desktop' 
.PP
Referenced by XcmColorServerCapabilities(), XcmeContext_Setup(), and XcmeContext_Setup2()\&.
.SS "#define XCM_COLOR_OUTPUTS   '_ICC_COLOR_OUTPUTS'"
The XCM_COLOR_OUTPUTS macro is attached to windows and specifies on which output the window should look correctly\&. The type is XA_STRING\&. 
.PP
Referenced by XcmeContext_Setup(), and XcmeContext_Setup2()\&.
.SS "#define XCM_COLOR_PROFILES   '_ICC_COLOR_PROFILES'"
The XCM_COLOR_PROFILES macro represents a atom name for under which \fBXcolorProfile\fP data can be stored\&. 
.PP
Referenced by XcmeContext_Setup(), XcmeContext_Setup2(), XcolorProfileDelete(), and XcolorProfileUpload()\&.
.SS "#define XCM_COLOR_REGIONS   '_ICC_COLOR_REGIONS'"
The XCM_COLOR_REGIONS macro represents a atom name for under which XcolorRegions data can be stored\&. 
.PP
Referenced by XcmeContext_Setup(), XcmeContext_Setup2(), XcmePrintWindowRegions(), xcmePrintWindowRegions(), XcolorRegionDelete(), XcolorRegionFetch(), and XcolorRegionInsert()\&.
.SS "#define XCM_DEVICE_PROFILE   '_ICC_DEVICE_PROFILE'"
The _ICC_DEVICE_PROFILE atom The atom will hold a native ICC profile with the exposed device characteristics at the compositing window manager level\&. The colour server shall if no _ICC_DEVICE_PROFILE(_xxx) is set, copy the _ICC_PROFILE(_xxx) profiles to each equivalent _ICC_DEVICE_PROFILE(_xxx) atom\&. The _ICC_PROFILE(_xxx) profiles shall be replaced by a sRGB ICC profile\&. The counting in the atoms (_xxx) name section follows the rules outlined in the ICC Profile in X recommendation\&. After finishing the session the the old state has to be recovered by copying any _ICC_DEVICE_PROFILE(_xxx) atoms content into the appropriate _ICC_PROFILE(_xxx) atoms and removing all _ICC_DEVICE_PROFILE(_xxx) atoms\&. The colour server must be aware about change property events indicating that a _ICC_PROFILE(_xxx) atom has changed by a external application and needs to move that profile to the appropriate _ICC_DEVICE_PROFILE(_xxx) atom and set the _ICC_PROFILE(_xxx) atom to sRGB as well\&. The modification of the _ICC_DEVICE_PROFILE(_xxx) atoms by external applications is undefined\&. 
.SH "Enumeration Type Documentation"
.PP 
.SS "anonymous enum"
The XCM_COLOR_SERVER_ enums describe colour server capabilities as specified in XCM_COLOR_DESKTOP atom\&. 
.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIXCM_COLOR_SERVER_REGIONS \fP\fP
_ICC_COLOR_REGIONS 
.TP
\fB\fIXCM_COLOR_SERVER_PROFILES \fP\fP
_ICC_COLOR_PROFILES 
.TP
\fB\fIXCM_COLOR_SERVER_DISPLAY_ADVANCED \fP\fP
_ICC_COLOR_DISPLAY_ADVANCED 
.TP
\fB\fIXCM_COLOR_SERVER_OUTPUTS \fP\fP
_ICC_COLOR_OUTPUTS 
.TP
\fB\fIXCM_COLOR_SERVER_MANAGEMENT \fP\fP
_ICC_COLOR_MANAGEMENT 
.SH "Function Documentation"
.PP 
.SS "int XcmColorServerCapabilities (Display *dpy)"

.PP
informs which colour server services are available Function XcmColorServerCapabilities Query the capabilities of a colour server\&. In case no colour server is running, the function should return zero\&. The return value consists of a bit mask of XCM_COLOR_SERVER_ properties from the XCM_COLOR_DESKTOP atom\&. 
.PP
References XCM_COLOR_DESKTOP, XCM_COLOR_SERVER_DISPLAY_ADVANCED, XCM_COLOR_SERVER_MANAGEMENT, XCM_COLOR_SERVER_PROFILES, and XCM_COLOR_SERVER_REGIONS\&.
.SS "int XcolorProfileDelete (Display *dpy, \fBXcolorProfile\fP *profile)"

.PP
Decreases the ref-count of the profile\&. Function XcolorProfileDelete You shouldn't use the profile anymore after this call because it could have been removed from the internal database in the compositing manager\&. If you do, nothing bad will happen, just that color management won't work on regions that use this profile\&. 
.PP
References XcolorProfile::length, and XCM_COLOR_PROFILES\&.
.SS "int XcolorProfileUpload (Display *dpy, \fBXcolorProfile\fP *profile)"

.PP
Uploads the profile into all screens of the display\&. Function XcolorProfileUpload Uploads the profile into all screens of the display\&. Profiles are ref-counted inside the compositing manager, so make sure to call \fBXcolorProfileDelete()\fP before your application exits or when you don't need the profile anymore\&. 
.PP
References XcolorProfile::length, and XCM_COLOR_PROFILES\&.
.SS "int XcolorRegionActivate (Display *dpy, Windowwin, unsigned longstart, unsigned longcount)"

.PP
Activates regions\&. Function XcolorRegionActivate Activates 'count' regions starting at positiong 'start' in the stack\&. Unlike the other functions it does not check whether 'start + count' extends beyond the stack end\&. To disable all regions pass zero to 'count'\&. 
.SS "int XcolorRegionDelete (Display *dpy, Windowwin, unsigned longstart, unsigned longcount)"

.PP
Deletes regions\&. Function XcolorRegionDelete Deletes 'count' regions starting at 'start' in the stack\&. If 'start + count' in beyond the stack end, nothing will be done and the function returns '-1'\&. 
.PP
References XCM_COLOR_REGIONS, and XcolorRegionFetch()\&.
.SS "\fBXcolorRegion\fP* XcolorRegionFetch (Display *dpy, Windowwin, unsigned long *nRegions)"

.PP
Fetches the existing regions\&. Function XcolorRegionFetch Fetches the existing regions and returns an array of '\fBXcolorRegion\fP'\&. After you're done, free the array using XFree()\&. The number of regions is put into 'nRegions'\&. 
.PP
References XCM_COLOR_REGIONS\&.
.PP
Referenced by XcmePrintWindowRegions(), xcmePrintWindowRegions(), XcolorRegionDelete(), and XcolorRegionInsert()\&.
.SS "int XcolorRegionInsert (Display *dpy, Windowwin, unsigned longpos, \fBXcolorRegion\fP *region, unsigned longnRegions)"

.PP
Inserts the supplied regions into the stack\&. Function XcolorRegionInsert Inserts the supplied regions into the stack at the position 'pos', shifting the existing profiles upwards in the stack\&. If 'pos' is beyond the stack end, nothing will be done and the function returns '-1'\&. 
.PP
References XCM_COLOR_REGIONS, and XcolorRegionFetch()\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for Xcm-0\&.5\&.4 from the source code\&.


ChangeLog
Version libXcm-0.5.3

Florian Höch (1):
	* [core]: Patch for transposed EDID red y bits

Kai-Uwe Behrmann (20):
	* [build]: remove all libtool files from RPM
	* [core]: fix conditional compiling of XcmDummy
	* [build]: fix spec packaging issues
	* [core]: call split libraries in libXcm
	* [build]: sync spec file with obs
	* [core]: add XcmDummy.c file
	* [build]: add xcm-edid xcm-x11 xcm-ddc pkg-config files
	* [build]: reorder building of libraries
	* [build]: split EDID parsing library out
	* [core]: XcmeContext_InLoop return meaningful value
	* [core]: Fix window name print
	* [conf]: update debian changelog
	* [docu]: update man pages
	* [core]: omit XmuClientWindow()
	* [conf]: bump version
	* [core]: initialy observe windows
	* [docu]: fix link
	* [build]: update autotools files
	* [build]: remove libtool
	* [build]: add missed automake script

Version libXcm-0.5.2

Jan Engelhardt (2):
	build: support for automake-1.12
	build: do not fail if AM_SILENT_RULES is not present

Jean-Sébastien Pédron (2):
	* [build]: support build dir != source dir
	* [build]: regen autotools files to correctly support FreeBSD 10

Kai-Uwe Behrmann (11):
	* [build]: rename FindXcm.cmake to XcmConfig.cmake
	* [conf]: install FindXcm.cmake
	* [core]: define XCM_COLOR_SERVER_MANAGEMENT enum
	* [docs]: describe ICM in X Color Management spec
	* [build]: fix debian install file names
	* [build]: add needed build requirements for debian
	* [build]: add requred dummy file
	* [build]: add make debian target
	* [docu]: mention opt out in X Color Management spec
	* [conf]: bump version
	* [build]: reset release version in RPM spec file

Version libXcm-0.5.1

Kai-Uwe Behrmann (4):
	* [core]: add XCM_ prefix to HAVE macros
	* [core]: simplify varg allocation for messages
	* [build]: depend on dist-bzip2 for make rpm target
	* [conf]: fix DDC code linking

Michael Weber (1):
	* [conf]: add --disable-x11 configure option


Version libXcm-0.5.0

Jan Engelhardt (7):
	* [conf]: allow building without X11
	* [build]: allow building without X11
	* [core]: resolve compiler warnings about incorrect code
	* [conf]: Xcm header files utilize header files from xproto (Xatom.h) and x11 (Xlib.h)
	* [core]: resolve compiler warnings about incorrect code
	* [conf]: switch to autotools
	* [conf]: Xcm header files utilize header files from xproto (Xatom.h) and x11 (Xlib.h)

Kai-Uwe Behrmann (75):
	* [build]: correct configure version macor in spec
	* [build]: update spec file for autotools changes
	* [core]: keep capabilities visible for non X11
	* [build]: add missed autotools files
	* [core]: skip "WARNING" for system message in xcmevents
	* [conf]: add xproto x11 conditionally to xcm.pc
	* [conf]: HAVE_X11 conditional for Xcm.h
	* [conf]: fix XcmEvents building
	* [conf]: copy build system
	* [core]: add XCM_COLOR_OUTPUTS to header
	* [docu]: adapt the X Color Management spec intro
	* [docu]: X Color Management specification 0.4
	* [conf]: fix Version field
	* [conf]: add Version field to pkg-config file
	* [conf]: fill all needed paths for pkg-config
	* [conf]: add config.h.in
	* [conf]: add build-aux files
	* [conf]: track essential build scripts and files
	* [conf]: remove old configuration files
	* [conf]: add status at end of configure run
	* [build]: add example files to package
	* [build]: add make help target
	* [build]: add make rpm and deb targets
	* [docu]: fix doxygen warning
	* [core]: use XcmVersion.h macros
	* [conf]: write HAVE_X11 to XcmVersion.h
	* [conf]: set fine grained package and library verions
	* [build]: pass libtool version to linker
	* [build]: readd missed files to tar ball
	* [conf]: adapt spec file template to autotools
	* [docu]: show macros for doxygen man pages
	* [conf]: add version header template
	* [conf]: add urls to configure.ac
	* [conf]: prepare autotools patch
	* [conf]: generalise Win32 detection
	* [conf]: check for MINGW32
	* [conf]: always use -fPIC
	* [core]: print ser_id
	* [core]: fix ser_id field to 32bit
	* [core]: add XcmColorServerCapabilities()
	* [core]: fix colorimetry name macros
	* [core]: fix string buffer allocation
	* [core]: fix key name swapping
	* [docu]: X Color Management v0.3 DRAFT2
	* [core]: adapt new XCM_COLOUR_DESKTOP_ADVANCED macro
	* [core]: detect _ICC_COLOR_DISPLAY_ADVANCED events
	* [docu]: add _ICC_COLOR_DISPLAY_ADVANCED + _ICC_COLOR_PROFILES
	* [core]: add a name to the XcmEvents observer window
	* [core]: add XcmeContext_Setup2 API
	* [core]: add SEC manufacturer
	* [core]: add GCM and LGE
	* [core]: be more strict about EDID primariy order
	* [core]: add LGD manufacturer
	* [conf]: bump to 0.5
	* [docu]: update to new XCM macro
	* [build]: fix spec file
	* [core]: use XCM_COLOR_ macros
	* [core]: device XCM atom strings
	* [docu]: correct the capabilities tagging XCM docu
	* [core]: rename net-color spec to X_Color_Management
	* [core]: export numbers as strings for JSON
	* [docu]: update autogenerated man pages
	* [exmpl]: use XcolorRegionDelete for 0 0 0 0 geometry
	* [core]: properly export XcmePrintWindowName
	* [build]: add client example make check target
	* [exmpl]: make profile selectable
	* [build]: add build line for color
	*  [exmpl]: compile color with C
	* [exmpl]: fix client
	* [build]: adapt spec file to openSUSE conventions
	* [exmpl]: add very simple makefile for color example
	* [core]: return Xorg error values
	* [exmpl]: update to actual library
	* [conf]: bump version
	* [core]: chose first text block for type

Version libXcm-0.4.2

Kai-Uwe Behrmann (11):
	* [core]: use only JSON array in xcmedid
	* [core]: skip the numbered JSON device level in xcmedid
	* [build]: default to bz2 compression for RPM
	* [build]: remove unused macros in RPM spec file
	* [core]: merge EDID week and year to EDID_date
	* [core]: fall back for EDID_model to model_id
	* [docu]: update XcmEdidParse man page
	* [core]: add AUO and LEN manufacturers
	* [core]: add XcmEdidPrintOpenIccJSON() API
	* [core]: exclude DDC from __FreeBSD__
	* [conf]: bump version


src/Xcm.c
/*  @file Xcm.c
 *
 *  libXcm  Xorg Colour Management
 *
 *  @par Copyright:
 *            2008 (C) Tomas Carnecky
 *            2008-2011 (C) Kai-Uwe Behrmann
 *
 *  @brief    X Color Management specification helpers
 *  @internal
 *  @author   Tomas Carnecky
 *            Kai-Uwe Behrmann <ku.b@gmx.de>
 *  @par License:
 *            MIT <http://www.opensource.org/licenses/mit-license.php>
 *  @since    2008/04/00
 */

#include <Xcm.h>
#include <stdio.h>

int XcolorProfileUpload(Display *dpy, XcolorProfile *profile)
{
	/* XcolorProfile::length is in network byte-order, swap it now */
	uint32_t length = htonl(profile->length);
        int i;

	Atom netColorProfiles = XInternAtom(dpy, XCM_COLOR_PROFILES, False);

	for (i = 0; i < ScreenCount(dpy); ++i) {
		XChangeProperty(dpy, XRootWindow(dpy, i), netColorProfiles, XA_CARDINAL, 8, PropModeAppend, (unsigned char *) profile, sizeof(XcolorProfile) + length);
	}

	return 0;
}

int XcolorProfileDelete(Display *dpy, XcolorProfile *profile)
{
	Atom netColorProfiles = XInternAtom(dpy, XCM_COLOR_PROFILES, False);
        int i;

	/* To delete a profile, send the header with a zero-length. */
	profile->length = 0;

	for (i = 0; i < ScreenCount(dpy); ++i) {
		XChangeProperty(dpy, XRootWindow(dpy, i), netColorProfiles, XA_CARDINAL, 8, PropModeAppend, (unsigned char *) profile, sizeof(XcolorProfile));
	}

	return 0;
}



int XcolorRegionInsert(Display *dpy, Window win, unsigned long pos, XcolorRegion *region, unsigned long nRegions)
{
	Atom netColorRegions = XInternAtom(dpy, XCM_COLOR_REGIONS, False);
	XcolorRegion *ptr;
	int result;

	unsigned long nRegs;
	XcolorRegion *reg = XcolorRegionFetch(dpy, win, &nRegs);

	/* Security check to ensure that the client doesn't try to insert the regions
	 * to a position beyond the stack end. */
	if (pos > nRegs) {
		XFree(reg);
		return -1;
	}

	ptr = calloc(sizeof(char), (nRegs + nRegions) * sizeof(XcolorRegion));
	if (ptr == NULL) {
		XFree(reg);
		return -1;
	}


	/* Make space for the new regions and copy them to the array. */
	if (nRegs)
	{
		memcpy(ptr, reg, nRegs * sizeof(XcolorRegion));
		memmove(ptr + pos + nRegs, ptr + pos, nRegions * sizeof(XcolorRegion));
	}
	memcpy(ptr + pos, region, nRegions * sizeof(XcolorRegion));

	result = !XChangeProperty(dpy, win, netColorRegions, XA_CARDINAL, 8, PropModeReplace, (unsigned char *) ptr, (nRegs + nRegions) * sizeof(XcolorRegion));

	if(reg)
		XFree(reg);
	free(ptr);

	return result;
}

XcolorRegion *XcolorRegionFetch(Display *dpy, Window win, unsigned long *nRegions)
{

	Atom actual, netColorRegions = XInternAtom(dpy, XCM_COLOR_REGIONS, False);

	unsigned long left, nBytes;
	unsigned char *data;
       
	int format, result = XGetWindowProperty(dpy, win, netColorRegions, 0, ~0, False, XA_CARDINAL, &actual, &format, &nBytes, &left, &data);

	*nRegions = 0;
	if (result != Success)
		return NULL;

	*nRegions = nBytes / sizeof(XcolorRegion);
	return (XcolorRegion *) data;
}


int XcolorRegionDelete(Display *dpy, Window win, unsigned long start, unsigned long count)
{
	Atom netColorRegions = XInternAtom(dpy, XCM_COLOR_REGIONS, False);
	int result;

	unsigned long nRegions;
	XcolorRegion *region = XcolorRegionFetch(dpy, win, &nRegions);

	/* Security check to ensure that the client doesn't try to delete regions
	 * beyond the stack end. */
	if (start + count > nRegions) {
		XFree(region);
		return -1;
	}

	/* Remove the regions and close the gap. */
	memmove(region + start, region + start + count, (nRegions - start - count) * sizeof(XcolorRegion));

  if(nRegions - count)
  	result = !XChangeProperty(dpy, win, netColorRegions, XA_CARDINAL, 8, PropModeReplace, (unsigned char *) region, (nRegions - count) * sizeof(XcolorRegion));
  else
    result = !XDeleteProperty( dpy, win, netColorRegions );

  XFree(region);


	return result;
}

int XcolorRegionActivate(Display *dpy, Window win, unsigned long start, unsigned long count)
{
	XWindowAttributes xwa;
	Status status;
	int result;

	/* Construct the XEvent. */
	XClientMessageEvent event;

	event.type = ClientMessage;
	event.window = win;
	event.message_type = XInternAtom(dpy, "_ICC_COLOR_MANAGEMENT", False);
	event.format = 32;

	event.data.l[0] = start;
	event.data.l[1] = count;

	/* The ClientMessage has to be sent to the root window. Find the root window
	 * of the screen containing 'win'. */
        status = XGetWindowAttributes(dpy, RootWindow(dpy, 0), &xwa);
	if (status == 0)
		return -1;

	/* Uhm, why ExposureMask? */
	result = XSendEvent(dpy, xwa.root, False, ExposureMask, (XEvent *) &event);

	return result;
}

static unsigned char * XcmFetchProperty(Display *dpy, Window w, Atom prop, Atom type, unsigned long *n, Bool del)
{
  Atom actual;
  int format;
  unsigned long left;
  unsigned char *data;
  int result;

  XFlush( dpy );

  result = XGetWindowProperty( dpy, w, prop, 0, ~0, del, type,
                               &actual, &format, n, &left, &data);
  if (result == Success)
    return data;

  return NULL;
}

int    XcmColorServerCapabilities    ( Display *dpy )
{
  int active = 0;
  unsigned long n = 0;
  unsigned char * data = 0;
  Atom iccColorDesktop = XInternAtom(dpy, XCM_COLOR_DESKTOP, False);

  data = XcmFetchProperty( dpy, RootWindow(dpy,0),
                           iccColorDesktop, XA_STRING, &n, False);
  if(data && n && strlen((char*)data))
  {
    int old_pid = 0;
    long atom_last_time = 0;
    char * atom_time_text = (char*)malloc(1024),
         * atom_colour_server_name = (char*)malloc(1024),
         * atom_capabilities_text = (char*)malloc(1024);

    atom_time_text[0]= atom_colour_server_name[0]= atom_capabilities_text[0]= 0;

    sscanf( (const char*)data, "%d %ld %s %s",
              &old_pid, &atom_last_time,
              atom_capabilities_text, atom_colour_server_name );
    if(atom_capabilities_text[0])
    {
      if(strstr(atom_capabilities_text, "|ICP|"))
        active |= XCM_COLOR_SERVER_PROFILES;
      if(strstr(atom_capabilities_text, "|ICR|"))
        active |= XCM_COLOR_SERVER_REGIONS;
      if(strstr(atom_capabilities_text, "|ICA|"))
        active |= XCM_COLOR_SERVER_DISPLAY_ADVANCED;
      if(strstr(atom_capabilities_text, "|ICM|"))
        active |= XCM_COLOR_SERVER_MANAGEMENT;
    }
    free(atom_capabilities_text);
    free(atom_time_text);
    free(atom_colour_server_name);
  }
  return active;
}

